ql/utilities/null.hpp

```diff
@@ -32,9 +32,9 @@
 
 namespace QuantLib {
 
-    //! template class providing a null value for a given type.
+    //! template function providing a null value for a given type.
     template <class Type>
-    class Null;
+    Type Null();
```

Comment:
This function prototype and all specializations should be made `constexpr` to allow it to be used at compile-time.

Reply:
This was done in the latest commit.

Reply:
The prototype itself should be left as non-`constexpr` because not all specializations support `constexpr` (namely the one for `Array`). Only the relevant specializations should be marked as `constexpr`. I had edited my original comment to clarify this - sorry for the confusion if there was any.

Reply:
In fact, in `null.hpp`, the declaration on top of the file is simply a declaration - and the function on the bottom is the generic definition of the template for any `T`, which is `constexpr`. We've just added a commit to clarify this (removing the declaration altogether). 

Unless we limit the generic implementation in `null.hpp` to specific types (e.g. POD types or arithmetic types or something like this with `enable_if`), your suggested approach would not work. The compiler produces an error that the declaration and definition is not matching because of the `constexpr`. 

The way it is now is actually standard compliant, as explicit specializations of a `constexpr` template are allowed to differ from the generic template with respect to `constexpr`. See Note 1 here: http://www.eel.is/c++draft/dcl.constexpr . However, GCC 4.8 does not seem to respect this, while all other compilers do. I see the following options to fix this:

1. Avoid `constexpr` in this function template for GCC 4.8 only, with some kind of `#ifdef`, so that people still using this old compiler get code that compiles ok (GCC 4.8 was released in 2013)
2. Don't make `Null<T>()` `constexpr` at all, given that it is not used in a compile-time context anywhere in the whole library at all at this point. In other uses, the compiler will optimise it out into a constant anyway in release mode. 

Note that if we went for the first option, we would also need to modify the codebase at the point of use in many places to fix the static analysis checks, i.e. change to `constexpr Real x = Null<Real>();`. However, this can create problems if a `Real` type is used that is not `double`. Therefore it may be better to supress these static analysis warnings and leave the code as is - or go for option 2.

Reply:
Removing the declaration altogether looks like the right choice to me. I have no further comments if the CI passes with this change.

FYI support for GCC 4.8 will be dropped in https://github.com/lballabio/QuantLib/pull/1324 which will be included in 1.28 sometime later this summer, so the less we have to worry about 4.8 between now and then the better.

Reply:
Ok - we have just committed this change.